substrate/client/src/block_builder.rs

```diff
@@ -23,12 +23,20 @@ use runtime_primitives::traits::{Header as HeaderT, Hash, Block as BlockT, One,
 use runtime_primitives::generic::BlockId;
 use {backend, error, Client, CallExecutor};
 use runtime_primitives::{ApplyResult, ApplyOutcome};
+use patricia_trie::NodeCodec;
+use primitives::{BlakeHasher, BlakeRlpCodec};
+use hashdb::Hasher;
+use rlp::Encodable;
 
 /// Utility for building new (valid) blocks from a stream of extrinsics.
-pub struct BlockBuilder<B, E, Block> where
-	B: backend::Backend<Block>,
-	E: CallExecutor<Block> + Clone,
+pub struct BlockBuilder<B, E, Block, H, C>
+where
+	B: backend::Backend<Block, H, C>,
+	E: CallExecutor<Block, H, C> + Clone,
 	Block: BlockT,
+	H: Hasher,
```

Comment:
reason for not integrating `H` and `C` into `BlockT`?

Reply:
Not sure I understand what you mean here. You mean something like this?

```
pub trait Block: Clone + Send + Sync + Codec + Eq + MaybeSerializeDebug + 'static {
	type Extrinsic: Member + Codec;
	type Header: Header<Hash=Self::Hash>;
	type Hash: Member + ::rstd::hash::Hash + Copy + MaybeDisplay + Default + SimpleBitOps + Codec + AsRef<[u8]>;
	type Hasher: Hasher<Out=<Self::Hash>::Output>;
	type NodeCodec: NodeCodec<Self::Hasher>;
…
```

There is some significant overlap between the `hashdb::Hasher` and substrate own `Hash` trait and when talking this over with @rphmeier his idea was to get this in first and refine/unify further down the line. Not sure if `Block` is the proper unification point but might very well be?